workbench-core/accounts/src/services/projectService.ts

```diff
@@ -127,91 +141,58 @@ export default class ProjectService {
-      key: { name: 'resourceType', value: this._resourceType },
-      index: 'getResourceByCreatedAt',
-      limit: pageSize,
-      filter: filterExpression,
-      values: expressionAttributeValues
-    };
-
-    let projects: Project[] = [];
-
-    let returnedCount = 0;
-    while (returnedCount < pageSize && returnedCount < projectIds.length) {
-      queryParams = addPaginationToken(paginationToken, queryParams);
-
-      const projectsResponse = await this._aws.helpers.ddb.query(queryParams).execute();
-
-      returnedCount += projectsResponse.Count!;
-
-      // nothing returned
-      if (projectsResponse.Items === undefined || _.isEmpty(projectsResponse.Items)) {
-        paginationToken = getPaginationToken(projectsResponse);
-        console.log(paginationToken);
-      }
-
-      // nothing else to get from DDB
-      else if (projectsResponse.LastEvaluatedKey === undefined) {
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
-        paginationToken = undefined;
+    const keys: { [key: string]: unknown }[] = projectIds.map((projectId) =>
+      buildDynamoDBPkSk(projectId, resourceTypeToKey.project)
+    );
+    const projectsResponse = (await this._aws.helpers.ddb.get(keys).execute()) as BatchGetItemCommandOutput;
+    if (projectsResponse.Responses) {
+      const projects: Project[] = projectsResponse.Responses[this._tableName].map((item) =>
+        this._formatFromDDB(item)
+      );
+      // apply sort or filter
+      let projectsOnPage: Project[] = projects;
+      if (filter) {
+        projectsOnPage = manualFilterProjects(filter, projects);
       }
-
-      // too little--while loop continues
-      else if (returnedCount < pageSize) {
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
-        paginationToken = getPaginationToken(projectsResponse);
+      if (sort) {
+        projectsOnPage = manualSortProjects(sort, projects);
+      } else if (filter === undefined) {
+        // default sort is by createdAt
+        projectsOnPage = manualSortProjects({ createdAt: 'asc' }, projects);
       }
-
-      // juuuust the right amount--getPaginationToken from response and exit loop
-      else if (returnedCount === pageSize) {
-        paginationToken = getPaginationToken(projectsResponse);
-        projects = projects.concat(projectsResponse.Items.map((item) => this._formatFromDDB(item)));
+      // build page and pagination token
+      if (paginationToken) {
+        const manualExclusiveStartKey = fromPaginationToken(paginationToken);
+        const exclusiveStartProjectId = manualExclusiveStartKey.pk.split('#')[1];
+        const exclusiveStartProject = projectsOnPage.find(
+          (project) => project.id === exclusiveStartProjectId
+        );
+        if (exclusiveStartProject === undefined) {
+          throw Boom.badImplementation('Something went wrong with listing projects');
```

Comment:
should this be more descriptive?

Reply:
This check would fail if the pagination token is invalid, if the pk is not formatted right from the parsed pagination token, or if the id from the pagination token doesn't exist in the list of projects. Do you have a suggestion for the error message?

Reply:
I do not have a suggestion for a more descriptive error message considering the conditions this would fail. IMO, when this fails, it is likely not something the user intentionally did and I have no suggestion or way to pinpoint exactly what went wrong. Hence, says something went wrong. That is why I asked for suggestions and explained the conditions in case anyone else had a better idea!

Reply:
It seems like all 3 options are related to an invalid pagination token. Maybe something like `Invalid pagination token`?

Reply:
I am good with that. Thanks for the suggestion!